clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java

```diff
@@ -0,0 +1,372 @@
+import java.util.Optional;
+import java.util.Queue;
+
+import static org.apache.kafka.clients.consumer.internals.AbstractFetch.requestMetadataUpdate;
+
+/**
+ * {@code FetchCollector} operates at the {@link RecordBatch} level, as that is what is stored in the
+ * {@link FetchBuffer}. Each {@link org.apache.kafka.common.record.Record} in the {@link RecordBatch} is converted
+ * to a {@link ConsumerRecord} and added to the returned {@link Fetch}.
+ *
+ * @param <K> Record key type
+ * @param <V> Record value type
+ */
+public class FetchCollector<K, V> {
+
+    private final Logger log;
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptions;
+    private final FetchConfig<K, V> fetchConfig;
+    private final FetchMetricsManager metricsManager;
+    private final Time time;
+
+    public FetchCollector(final LogContext logContext,
+                          final ConsumerMetadata metadata,
+                          final SubscriptionState subscriptions,
+                          final FetchConfig<K, V> fetchConfig,
+                          final FetchMetricsManager metricsManager,
+                          final Time time) {
+        this.log = logContext.logger(FetchCollector.class);
+        this.metadata = metadata;
+        this.subscriptions = subscriptions;
+        this.fetchConfig = fetchConfig;
+        this.metricsManager = metricsManager;
+        this.time = time;
+    }
+
+    /**
+     * Return the fetched {@link ConsumerRecord records}, empty the {@link FetchBuffer record buffer}, and
+     * update the consumed position.
+     *
+     * </p>
+     *
+     * NOTE: returning an {@link Fetch#empty() empty} fetch guarantees the consumed position is not updated.
+     *
+     * @param fetchBuffer {@link FetchBuffer} from which to retrieve the {@link ConsumerRecord records}
+     *
+     * @return A {@link Fetch} for the requested partitions
+     * @throws OffsetOutOfRangeException If there is OffsetOutOfRange error in fetchResponse and
+     *         the defaultResetPolicy is NONE
+     * @throws TopicAuthorizationException If there is TopicAuthorization error in fetchResponse.
+     */
+    public Fetch<K, V> collectFetch(final FetchBuffer fetchBuffer) {
+        final Fetch<K, V> fetch = Fetch.empty();
+        final Queue<CompletedFetch> pausedCompletedFetches = new ArrayDeque<>();
+        int recordsRemaining = fetchConfig.maxPollRecords;
+
+        try {
+            while (recordsRemaining > 0) {
+                final CompletedFetch nextInLineFetch = fetchBuffer.nextInLineFetch();
+
+                if (nextInLineFetch == null || nextInLineFetch.isConsumed()) {
+                    final CompletedFetch records = fetchBuffer.peek();
+
+                    if (records == null)
+                        break;
+
+                    if (!records.isInitialized()) {
+                        try {
+                            fetchBuffer.setNextInLineFetch(initialize(records));
```

Comment:
I absolutely agree that it's awkward. This is essentially copy-and-paste from the existing implementation and I wanted to change things as little as possible. I'll take another look and see if I can make it more intuitive and easy to follow.

Reply:
I'm gonna call a time out on this because I've spent a couple of hours looking into this and am getting nowhere.

My attempts are either a) subtly wrong because all tests fail!, or b) no more clear than the code is now.

It's an irritatingly impregnable bit of code, I'll admit. Do you mind if I file a Jira and come back to this a bit later?

LMK. Thanks!